Document CMPv2 (#27915)
hashicorp/vault: website/content/docs/secrets/pki/cmpv2.mdx

---
layout: docs
page_title: Certificate Management Protocol v2 (CMPv2) within Vault | PKI - Secrets Engines

> description: An overview of the Enrollment over Secure Transport protocol implementation within Vault.

Author: Just a bug

---

# PKI secrets engine - Certificate Management Protocol v2 (CMPv2) <EnterpriseAlert inline="true" />

This document covers configuration and limitations of Vault's PKI Secrets Engine

> implementation of the [CMPv2 protocol](https://datatracker.ietf.org/doc/html/rfc4210) <EnterpriseAlert inline="true" />.

Author: The problem is the docs aren't structured to have more than 2 views on any topic, a descriptive section (which is this doc), and an API doc.  How-to docs have traditionally been in learn guides if they're more step by step than what we can do here.  I will rephrase that introductory statement though.

## What is Certificate Management Protocol v2 (CMPv2)?

The CMP protocol is an IETF standardized protocol, [RFC 4210](https://datatracker.ietf.org/doc/html/rfc4210),
that allows clients to acquire client certificates and their associated Certificate
Authority (CA) certficates.

## Enabling CMPv2 support on a Vault PKI  mount

To configure an existing mount to serve EST clients, the following steps are
required, which are broken down into three main categories:

 1. [Authentication mechanisms](#configuring-est-authentication)
 1. [Updating PKI tunable parameters](#updating-the-pki-mount-tunable-parameters)
 1. [PKI CMPv2 configuration](#pki-cmpv2-configuration)

### Configuring CMPv2 Authentication

At this time, Vault's implementation of CMPv2 supports only 
[Certificate TLS authentication](/vault/docs/auth/cert), where clients proof
of posession of a TLS client certificate authenticates them to Vault.

Authentication leverages a separate Vault authentication
mount, within the same namespace, to validate the client provided credentials
along with the client's ACL policy to enforce. 

> For proper accounting, a mount supporting CMPv2 authentication should be
> dedicated to this purpose, not shared with other workflows.  In other words,
> create a new certificate auth mount for CMPv2 even if you already have one 
> another in use for other purposes.
> 
> When setting up the authentication mount for CMPv2 clients, the token type must
> be configured to return [batch tokens](/vault/docs/concepts/tokens#batch-tokens).
> Batch tokens are required to avoid an excessive amount of leases being generated
> and persisted as every CMPv2 incoming request needs to be authenticated.

Reviewer: ```suggestion

- To ensure proper accounting and auditing, do not share the CMPv2 mount with
  other workflows. Always create a dedicated certificate authentication mount
 for CMPv2.
- Configure the token type for your CMPv2 authentication mount to return
  [batch tokens](/vault/docs/concepts/tokens#batch-tokens) to avoid excessive
  lease generation and persistance.
```

This feels like best practice recommendations rather than education (overview)
or instructional (how-to) information. If we have enough recommendations to
justify a separate page, I recommend moving it to a dedicated document. But if
we only have a few recs, I'd recommend creating a dedicated "Best practices"
section in the overview an collecting the recommendations there. (I've styled the
suggestion assuming the latter).

Style correction: write in active voice.

> The path within an ACL policy must match the `cmp` path underneath the 
> PKI mount.  The path to use can be the default `cmp` path or a role based one.

Reviewer: ```suggestion
ACL policy paths for CMPv2 clients must match to an appropriate `cmp`
path for your PKI mount. You can use the default `cmp` path or a
role-based path.
```

I tried to edit these sentences to use active voice, but I'm not entirely sure what the sentences were trying to say, so I may have inadvertently changed the meaning.

Style correction: write in active voice

> If using the `sign-verbatim` as a path policy, the following
> ACL policy will allow an authenticated client access the required PKI CMP path.

Reviewer: ```suggestion
For example, if you use the `sign-verbatim` as a path policy, you can define
an ACL policy using the default PKI CMP path, `pki/cmp`:
```

Style correction: write in active voice, speak directly to the reader with "you" for instructions.

```
path “pki/cmp” {
  capabilities=[“update”, “create”]
}
```

> For a role base path policy, this sample policy can be used

Reviewer: ```suggestion
Or a role-based path:
```

```
path “pki/roles/my-role-name/cmp” {
  capabilities=[“update”, “create”]
}

> ```

Reviewer: This is definitely how-to content, not overview content

#### Updating the PKI mount tunable parameters

> Once the authentication mount has been created and configured, the authentication mount's accessor
> will need to be captured and added within the PKI mount's [delegated auth accessors](/vault/api-docs/system/mounts#delegated_auth_accessors).

Reviewer: ```suggestion
Once you create and configure the authentication mount, you need to capture and add the relevant accessor to the list of [delegated auth accessors](/vault/api-docs/system/mounts#delegated_auth_accessors) for the PKI mount.
```

Rather than linking out to an API document, we should document the actual steps here, in the how-to guide and provide the API, GUI, and CLI commands appropriate for that step.

Style correction: write in active voice

> To get an authentication mount's accessor field, the following command can be used.

Reviewer: How do these instructions relate to the previous instrucitons?

```shell-session
$ vault read -field=accessor sys/auth/auth/cert
```

For CMP to work within certain clients, a few response headers need to be explicitly allowed
along with configuring the list of accessors the mount can delegate authentication towards.
The following will grant the required response headers, you will need to replace the values for the `delegated-auth-accessors`

> to match your values.

Reviewer: This is written like it's, potentially, an optional step since it's only "certain clients" that need this configuration step. What are the "certain clients" in these cases? As an operator, how do I know if I actually need to do this?

```shell-session
$ vault secrets tune \
  -allowed-response-headers="Content-Transfer-Encoding" \
  -allowed-response-headers="Content-Length" \
  -allowed-response-headers="WWW-Authenticate" \
  -delegated-auth-accessors="auth_cert_4088ac2d" \
  pki
```

#### Enabling CMPv2

> Enabling CMP is a matter of writing to the `config/cmp` endpoint, to set it
> enabled and configure default issuance policy and authentication.

Reviewer: ```suggestion
To enable CMPv2, configure a default issuance policy and authenticator for the `cmp` endpoint, of your PKI plugin. For example:
```

```shell-session
vault write pki/config/cmp -<<EOC
{
  "enabled": true,
  "authenticators": {
    "cert": {
      "accessor": "auth_cert_4088ac2d"
    }
  }
}
EOC
```

> Of course, substituting your own role and accessor values.  After this, the
> CMP endpoints will be available to handle client requests, authenticated with the 
> previously configured Cert Auth method.

Reviewer: ```suggestion
```
The fact that we're enabling the endpoint already suggests that it would be available, so I'm not sure these sentences are needed.

> ## Limitations

Reviewer: So far, this is the only overview-related content besides the best practices. I'd suggest restructuring the overview along the lines of:
```
# Certificate Management Protocol v2 (CMPv2) in the PKI plugin

@include 'alerts/enterprise-only.mdx' (or enterprice-and-hcp.mdx if HCP also supports it)

The PKI secrets engine supports [CMPv2](https://datatracker.ietf.org/doc/html/rfc4210)  for authentication workflows.

## What is CMPv2?

{GENERAL_OVERVIEW_CONTENT}

## Limitations

{LIMITATIONS_CONTENT}

## Best practices

{BEST_PRACTICES_CONTENT}
```

It may be a much shorter overview, but that will make it easier to scan and moving the how-to content to its own page will make the instructions easier to parse and follow

> The initial release of CMPv2 support is intentionally limited to a subset of the
> protocol, covering Initialization, Certification, and Key Update, over HTTP.
> In particular, the following are not yet supported:

Reviewer: ```suggestion
CMPv2 is currently limited to a subset of the protocol over HTTP:

- Initialization
- Certification
- Key Update

The following elements of CMPv2 are not supported:
```

 * Basic authentication scheme using PasswordBasedMac
 * Revocation
 * CRL fetching via CMP itself.
 * CA creation/update operations.

> In addition, CMPv2 will not yet be integrated with these existing Vault PKI features:

Reviewer: ```suggestion
In addition, CMPv2 is not supported with the following Vault PKI features:
```

 
 * Certificate Metadata - CMPv2 has no means of providing metadata.
 * Certificate Issuance External Policy Service [(CIEPS)](/vault/docs/concepts/pki/cieps) - 

>    May be implemented in the future.

Reviewer: I suggest either making this a table and providing the reasoning for both items, or leaving it a list and just noting the affected features without noting a reason. Saying something "may be implemented in the future" isn't helpful to the reader.

---
layout: docs
page_title: CMPv2 within Vault | PKI - Secrets Engines
description: An overview of the Certificate Management Protocol (v2) implementation within Vault.
---

# PKI secrets engine - Certificate Management Protocol v2 (CMPv2) <EnterpriseAlert inline="true" />

This document summarizes Vault's PKI Secrets Engine
implementation of the [CMPv2 protocol](https://datatracker.ietf.org/doc/html/rfc4210) <EnterpriseAlert inline="true" />,
its configuration, and limitations.

## What is Certificate Management Protocol v2 (CMPv2)?

The CMP protocol is an IETF standardized protocol, [RFC 4210](https://datatracker.ietf.org/doc/html/rfc4210),
that allows clients to acquire client certificates and their associated Certificate
Authority (CA) certficates.

## Enabling CMPv2 support on a Vault PKI  mount

> To configure an existing mount to serve EST clients, the following steps are

Reviewer 2: Need to change `EST` to `CMPv2` here. 

Also in the link in line 24.

to match your values.

```shell-session
$ vault secrets tune \
  -allowed-response-headers="Content-Transfer-Encoding" \
  -allowed-response-headers="Content-Length" \
  -allowed-response-headers="WWW-Authenticate" \
  -delegated-auth-accessors="auth_cert_4088ac2d" \
  pki
```

#### Enabling CMPv2

Enabling CMP is a matter of writing to the `config/cmp` endpoint, to set it
enabled and configure default path policy and authentication.

```shell-session
vault write pki/config/cmp -<<EOC
{
  "enabled": true,
  "default_path_policy": "role:example-role",
  "authenticators": {
    "cert": {
      "accessor": "auth_cert_4088ac2d"
    }
  }
}
EOC
```

Of course, substituting your own role and accessor values.  After this, the
CMP endpoints will be able to handle client requests, authenticated with the 
previously configured Cert Auth method.

## Limitations

CMPv2 is currently limited to a subset of the protocol over HTTP:

- Initialization
- Certification
- Key Update

The following elements of CMPv2 are not supported:

 * Basic authentication scheme using PasswordBasedMac
 * Revocation
 * CRL fetching via CMP itself.
 * CA creation/update operations.

> <<<<<<< HEAD

Reviewer 2: Need to clean up the merge conflict markers.
